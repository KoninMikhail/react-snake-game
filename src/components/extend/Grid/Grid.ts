import styled from 'styled-components';
import { gap } from '../../../styles/variables';

/**
 * @typedef IGridProps
 * @property {string=} columns - The columns template (css: grid-template-columns);
 * @property {string=} rows - The rows template (css: grid-template-rows);
 * @property {string=} gap - The gap size (css: gap);
 * @property {string=} align - Align items (css: align-items);
 * @property {string=} justify- Justify (css: justify-content);
 */
interface IGridProps {
  columns?: string;
  rows?: string;
  gap?: string;
  justify?: string;
  align?: string;
}

/**
 * Grid
 *
 * @description Use for create a grid layout on the page;
 *
 * @param {IGridProps} - The grid properties
 * @returns {JSX.Element}
 */
export const Grid = styled.div<IGridProps>`
  display: grid;
  position: relative;
  grid-template-columns: ${(props) => props?.columns || '1fr'};
  grid-template-rows: ${(props) => props?.rows || '1fr'};
  grid-gap: ${(props) => props?.gap || gap};
  justify-content: ${(props) => props?.justify || 'start'};
  align-items: ${(props) => props?.align || 'normal'};
`;
