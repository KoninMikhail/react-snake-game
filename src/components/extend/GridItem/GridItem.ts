import styled from 'styled-components';

/**
 * @typedef IGridItemProps
 *
 * @property {string=} gridColumn - Columns size (css: grid-column);
 * @property {string=} gridRow - Row size (css: grid-row);
 * @property {string=} align - Align in grid item (css: justify-self);
 * @property {string=} justify- Justify in grid item (css: align-self);
 */

interface IGridItemProps {
  gridColumn?: string;
  gridRow?: string;
  justify?: string;
  align?: string;
}

/**
 * GridItem
 *
 * @description Use for creation item in the grid
 *
 * @memberOf Grid
 * @param {IGridItemProps} - The grid item properties
 * @returns {JSX.Element}
 */

export const GridItem = styled.div<IGridItemProps>`
  display: block;
  position: relative;
  grid-column: ${(props) => props?.gridColumn};
  grid-row: ${(props) => props?.gridRow || 'auto'};
  justify-self: ${(props) => props?.justify || 'auto'};
  align-self: ${(props) => props?.align || 'auto'};
`;
