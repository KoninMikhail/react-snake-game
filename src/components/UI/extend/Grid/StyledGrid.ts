import styled from 'styled-components';
import { gap } from '../../../../styles/variables';
import { IGridProps } from './Grid';

export const StyledGrid = styled.div<IGridProps>`
  display: grid;
  position: relative;
  grid-template-columns: ${(props) => props?.columns || '1fr'};
  grid-template-rows: ${(props) => props?.rows || '1fr'};
  grid-gap: ${(props) => props?.gap || gap};
  justify-content: ${(props) => props?.justify || 'start'};
  align-items: ${(props) => props?.align || 'normal'};
`;
