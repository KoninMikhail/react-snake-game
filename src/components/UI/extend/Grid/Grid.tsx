import { StyledGrid } from './StyledGrid';
import { IGridItemProps } from '../GridItem/GridItem';
import { ReactElement } from 'react';
import { CSSProperties } from 'styled-components';

/**
 * @typedef IGridProps
 * @property {string=} columns - The columns template (css: grid-template-columns);
 * @property {string=} rows - The rows template (css: grid-template-rows);
 * @property {string=} gap - The gap size (css: gap);
 * @property {string=} align - Align items (css: align-items);
 * @property {string=} justify- Justify (css: justify-content);
 */
export interface IGridProps {
  columns?: string;
  rows?: string;
  gap?: string;
  justify?: string;
  align?: string;
  children: ReactElement<IGridItemProps> | Array<ReactElement<IGridItemProps>>;
  style?: CSSProperties;
}

/**
 * StyledGrid
 *
 * Use for create a grid layout on the page;
 *
 * @param {IGridProps} - The grid properties
 * @returns {JSX.Element}
 */

export const Grid = ({ children, ...rest }: IGridProps): JSX.Element => (
  <StyledGrid {...rest}>{children}</StyledGrid>
);
