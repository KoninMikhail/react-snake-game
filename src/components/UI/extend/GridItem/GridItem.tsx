import { CSSProperties } from 'styled-components';
import { StyledGridItem } from './StyledGridItem';

/**
 * @typedef IGridItemProps
 *
 * @property {string=} gridColumn - Columns size (css: grid-column);
 * @property {string=} gridRow - Row size (css: grid-row);
 * @property {string=} align - Align in grid item (css: justify-self);
 * @property {string=} justify- Justify in grid item (css: align-self);
 */

export interface IGridItemProps {
  gridColumn?: string;
  gridRow?: string;
  justify?: string;
  align?: string;
  children: JSX.Element | JSX.Element[] | string;
  style?: CSSProperties;
}

/**
 * GridItem
 *
 * Use for creation item in the grid
 *
 * @memberOf Grid
 * @param {IGridItemProps} props - The grid item properties
 * @returns {JSX.Element}
 */

export const GridItem = ({ children, ...rest }: IGridItemProps): JSX.Element => (
  <StyledGridItem {...rest}>{children}</StyledGridItem>
);
