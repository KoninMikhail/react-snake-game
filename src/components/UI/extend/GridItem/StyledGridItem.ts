import styled from 'styled-components';
import { IGridItemProps } from './GridItem';

export const StyledGridItem = styled.div<IGridItemProps>`
  display: block;
  position: relative;
  grid-column: ${(props) => props?.gridColumn || 'auto'};
  grid-row: ${(props) => props?.gridRow || 'auto'};
  justify-self: ${(props) => props?.justify || 'auto'};
  align-self: ${(props) => props?.align || 'auto'};
`;
