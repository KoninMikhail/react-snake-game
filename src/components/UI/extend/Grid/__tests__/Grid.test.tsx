import { render, screen } from '@testing-library/react';
import { Grid } from '../Grid';
import { GridItem } from '../../GridItem/GridItem';

describe('Grid', () => {
  const gridProps = {
    justify: 'center',
    align: 'center',
    columns: 'auto 1fr auto',
    rows: 'auto 55px',
    gap: '15px',
  };
  const gridTestId = 'gridParentElementTest';

  it('correctly render grid', () => {
    render(
      <Grid {...gridProps} data-testid={gridTestId}>
        <GridItem>{'item'}</GridItem>
      </Grid>
    );

    const gridElement = screen.getByTestId(gridTestId);

    expect(gridElement).toHaveStyleRule('display', 'grid');
    expect(gridElement).toHaveStyleRule('grid-template-columns', 'auto 1fr auto');
    expect(gridElement).toHaveStyleRule('grid-template-rows', 'auto 55px');
    expect(gridElement).toHaveStyleRule('grid-gap', '15px');
    expect(gridElement).toHaveStyleRule('justify-content', 'center');
    expect(gridElement).toHaveStyleRule('align-items', 'center');
  });
  it('equal snapshot', () => {
    render(
      <Grid {...gridProps} data-testid={gridTestId}>
        <GridItem>{'item'}</GridItem>
      </Grid>
    );

    const gridElement = screen.getByTestId(gridTestId);

    expect(gridElement).toMatchSnapshot();
  });
});
