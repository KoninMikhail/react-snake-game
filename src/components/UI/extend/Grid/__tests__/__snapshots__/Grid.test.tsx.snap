// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Grid equal snapshot 1`] = `
.c0 {
  display: grid;
  position: relative;
  grid-template-columns: auto 1fr auto;
  grid-template-rows: auto 55px;
  grid-gap: 15px;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c1 {
  display: block;
  position: relative;
  grid-column: auto;
  grid-row: auto;
  justify-self: auto;
  -webkit-align-self: auto;
  -ms-flex-item-align: auto;
  align-self: auto;
}

<div
  class="c0"
  data-testid="gridParentElementTest"
>
  <div
    class="c1"
  >
    item
  </div>
</div>
`;
